### ISSUE TITLE
Logind Support

### ISSUE THREAD

**Issue by Stebalien** [dunst-project/dunst#311]
It would be great if dunst could use logind to:

1. Track/find the primary graphical session instead of relying on the `DISPLAY` environment variable (for users who run dunst via systemd).
2. Suspend itself when the primary graphical session is locked (using the per-session `LockedHint` dbus property).

**Comment by tsipinakis** [comment]
I'm a bit hesitant to support this but I'm open to debate.

> Track/find the primary graphical session instead of relying on the `DISPLAY` environment variable (for users who run dunst via systemd).

What is the usecase for this? Dunst is usually started by a dbus activation when a notification is sent. I would assume that if a graphical session is running, `DISPLAY` would also be set.

> Suspend itself when the primary graphical session is locked (using the per-session LockedHint dbus property).

This might not always be the expected behaviour. Some users might expect that irrelevant notifications(i.e. 'currently playing' messages from media players) expire even if the screen is locked.

**Comment by Stebalien** [comment]
> What is the usecase for this? Dunst is usually started via a dbus activation when a notification is sent. I would assume that if a graphical session is running, DISPLAY would also be set.

On systems with systemd, dbus is usually run as a user service, not under a session, so it doesn't have access to the session's environment. The current dunst systemd service works around this by setting `DISPLAY=:0`. However, that's kind of a hack.

>  This might not always be the expected behaviour. Some users might expect that irrelevant notifications (i.e. 'currently playing' messages from media players) expire even if the screen is locked.

This could be configurable (one could, e.g., use urgency). However, my *primary* reason is that dunst notifications keep getting drawn on-top of my screen locker (i3lock).

**Comment by tsipinakis** [comment]
> On systems with systemd, dbus is usually run as a user service, not under a session, so it doesn't have access to the session's environment. The current dunst systemd service works around this by setting DISPLAY=:0. However, that's kind of a hack.

I don't see a way to do this without polluting the code with systemd-specific checks like this. I feel like this is a wrong design choice on systemds part, dunst shouldn't have to look in multiple places to find which display to use.

> This could be configurable (one could, e.g., use urgency). However, my primary reason is that dunst notifications keep getting drawn on-top of my screen locker (i3lock).

This is a case of trying to solve a bug by adding unnecessary complexity. Dunst shouldn't under any circumstances draw above the screen locker. Can you file this as a separate issue with any details you have to reproduce this? Bizarrely this is kind-of the opposite of #225.

**Comment by Stebalien** [comment]
> I don't see a way to do this without polluting the code with systemd-specific checks like this. I feel like this is a wrong design choice on systemds part, we shouldn't have to look in multiple places to find which display to use.

For better or worse, systemd decided that there should be one dbus instance per user, not per session. This means one dunst service per user, outside the session.

Unfortunately, after thinking how this would be implemented, I tend to agree. You'd need to be able to handle displays coming and going.

A better solution might be to fix this in the systemd unit by removing the `DISPLAY=:0` setting and adding:

```
[Unit]
...
After=graphical-session-pre.target
# Needs the DISPLAY environment variable
Requisite=graphical-session.target
PartOf=graphical=session.target

...

[Install]
WantedBy=graphical-session.target
```

Proper graphical sessions are supposed set the `DISPLAY` environment variable in systemd itself before starting `graphical-session.target` so this should work.

> This is a case of trying to solve a bug by adding unnecessary complexity.

You're probably right. After looking around, this looks like a bug in i3lock (https://github.com/i3/i3lock/issues/22).
